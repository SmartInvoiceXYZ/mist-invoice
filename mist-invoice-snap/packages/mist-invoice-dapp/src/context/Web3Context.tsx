import { SafeAppWeb3Modal as Web3Modal } from '@gnosis.pm/safe-apps-web3modal';
import WalletConnectProvider from '@walletconnect/web3-provider';
import { BrowserProvider, Eip1193Provider } from 'ethers';
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react';
import Web3, { EthExecutionAPI, Web3BaseProvider } from 'web3';

import { theme } from '../theme';
import { getRpcUrl, logError, ChainId, SUPPORTED_NETWORKS } from '../utils';

const providerOptions = {
  walletconnect: {
    package: WalletConnectProvider,
    options: {
      rpc: {
        1: getRpcUrl(1),
        4: getRpcUrl(4),
        42: getRpcUrl(42),
        100: getRpcUrl(100),
        137: getRpcUrl(137),
      },
    },
  },
};

type Web3ContextType = {
  loading: boolean;
  account?: string;
  provider?: BrowserProvider;
  chainId?: ChainId;
  connectAccount: () => Promise<void>;
  disconnect: () => Promise<void>;
};

type Web3ProviderState = {
  account?: string;
  provider?: BrowserProvider;
  chainId?: ChainId;
};

export const Web3Context = createContext<Web3ContextType>({
  loading: true,
  connectAccount: async () => {},
  disconnect: async () => {},
});
export const useWeb3 = () => useContext(Web3Context);

export const Web3ContextProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const [loading, setLoading] = useState(true);
  const [{ account, provider, chainId }, setWeb3] = useState<Web3ProviderState>(
    {},
  );
  const [web3Modal, setWeb3Modal] = useState<Web3Modal>();

  const setWeb3Provider = async (
    prov: Web3BaseProvider<EthExecutionAPI>,
    initialCall = false,
  ) => {
    if (prov) {
      const web3Provider = new Web3(prov);
      const gotProvider = new BrowserProvider(
        web3Provider.currentProvider as Eip1193Provider,
      );
      const network = await gotProvider.getNetwork();
      const gotChainId = Number(network.chainId) as ChainId;
      if (initialCall) {
        const signer = await gotProvider.getSigner();
        const gotAccount = await signer.getAddress();
        setWeb3((_provider: Web3ProviderState) => ({
          ..._provider,
          provider: gotProvider,
          chainId: gotChainId,
          account: gotAccount,
        }));
      } else {
        setWeb3((_provider: Web3ProviderState) => ({
          ..._provider,
          provider: gotProvider,
          chainId: gotChainId,
        }));
      }
    }
  };

  useEffect(() => {
    if (chainId && SUPPORTED_NETWORKS.indexOf(chainId as ChainId) === -1) {
      // TODO show error alert that invalid network is connected
    }
  }, [chainId]);

  const connectWeb3 = useCallback(async () => {
    try {
      if (!web3Modal) return;

      setLoading(true);
      const modalProvider = await web3Modal.requestProvider();

      await setWeb3Provider(modalProvider, true);

      const isGnosisSafe = !!modalProvider.safe;

      if (!isGnosisSafe) {
        modalProvider.on('accountsChanged', (accounts: string[]) => {
          setWeb3((_provider: Web3ProviderState) => ({
            ..._provider,
            account: accounts[0],
          }));
        });
        modalProvider.on('chainChanged', () => {
          setWeb3Provider(modalProvider);
        });
      }
    } catch (web3ModalError) {
      logError({ web3ModalError });
      throw web3ModalError;
    } finally {
      setLoading(false);
    }
  }, [web3Modal]);

  const disconnect = useCallback(async () => {
    if (!web3Modal) return;
    web3Modal.clearCachedProvider();
    setWeb3({});
  }, [web3Modal]);

  useEffect(() => {
    setWeb3Modal(
      new Web3Modal({
        cacheProvider: true,
        providerOptions,
        theme: {
          background: theme.colors.background,
          border: theme.colors.transparent,
          main: theme.colors.red[500],
          secondary: theme.colors.white,
          hover: theme.colors.black30,
        },
      }),
    );
  }, []);

  useEffect(() => {
    if (window.ethereum) {
      window.ethereum.autoRefreshOnNetworkChange = false;
    }
    (async function load() {
      if (web3Modal?.cachedProvider) {
        connectWeb3();
      } else {
        setLoading(false);
      }
    })();
  }, [connectWeb3, web3Modal]);

  return (
    <Web3Context.Provider
      value={{
        loading,
        account,
        provider,
        chainId,
        connectAccount: connectWeb3,
        disconnect,
      }}
    >
      {children}
    </Web3Context.Provider>
  );
};
