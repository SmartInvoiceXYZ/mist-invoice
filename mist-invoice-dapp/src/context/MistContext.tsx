import React, { createContext, useContext, useEffect, useState } from "react";

import { StandardMerkleTree } from "@openzeppelin/merkle-tree";
import skiffCrypto from "@skiff-org/skiff-crypto";
import { BytesLike } from "ethers";
import randombytes from "randombytes";
import { useIndexedDB } from "react-indexed-db-hook";
import { logDebug, logError } from "../utils";
import { Web3Context } from "./Web3Context";

export type MistData = {
  merkleRoot?: BytesLike;
  clientRandom: BytesLike;
  providerRandom: BytesLike;
  clientKey: BytesLike;
  providerKey: BytesLike;
};

export type MistContextType = {
  data?: MistData;
  loading: boolean;
  getClientProof: () => BytesLike[] | undefined;
};

export const MistContext = createContext<MistContextType>({
  getClientProof: () => undefined,
  loading: true
});

export const MistContextProvider: React.FC<React.PropsWithChildren> = ({
  children
}) => {
  const { getByID, add } = useIndexedDB(process.env.MIST_DB_NAME || "");
  const { account } = useContext(Web3Context);
  const [data, setData] = useState<MistData>();
  const [tree, setTree] =
    useState<StandardMerkleTree<(BytesLike | undefined)[]>>();
  const [loading, setLoading] = useState(false);

  const providerAddress = process.env.MIST_PROVIDER_ADDRESS || "";

  useEffect(() => {
    if (!account || !providerAddress || !data) return;
    const values = [
      [account, data.clientRandom],
      [providerAddress, data.providerRandom]
    ];

    const _tree = StandardMerkleTree.of(values, ["address", "bytes32"]);
    setTree(_tree);
    setData({
      ...data,
      merkleRoot: _tree.root
    });
  }, [account, providerAddress, data]);

  const generate = () => {
    const clientRandom = randombytes(32);
    const providerRandom = randombytes(32);
    const clientKey = skiffCrypto.generateSymmetricKey();
    const providerKey = skiffCrypto.generateSymmetricKey();

    const generated = {
      clientRandom,
      providerRandom,
      clientKey,
      providerKey
    };

    setData(generated);

    return generated;
  };

  useEffect(() => {
    if (account) {
      setLoading(true);
      try {
        getByID<MistData>(account).then((res) => {
          if (res) {
            logDebug("Using existing data");
            setData(res);
          } else {
            logDebug("Generating new data");
            res = generate();
            add(res, account).then(() => {
              setData(res);
            });
          }
        });
      } catch (e) {
        logError(e);
      } finally {
        setLoading(false);
      }
    } else {
      setData(undefined);
    }
  }, [account]);

  const getClientProof = () => {
    if (account && tree && data) {
      const proof = tree.getProof([account, data.clientRandom]);
      return proof;
    }
    return undefined;
  };

  return (
    <MistContext.Provider value={{ data, loading, getClientProof }}>
      {children}
    </MistContext.Provider>
  );
};
