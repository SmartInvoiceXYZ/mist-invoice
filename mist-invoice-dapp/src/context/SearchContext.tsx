import React, { createContext, useContext, useEffect, useState } from "react";

import { search } from "../graphql";
import { ChainId, logError } from "../utils";
import { Web3Context } from "./Web3Context";

export type SearchContextType = {};

export const SearchContext = createContext<SearchContextType>({});

export const SearchContextProvider: React.FC<React.PropsWithChildren<{}>> = ({
  children,
}) => {
  const { chainId } = useContext(Web3Context);
  const [fetching, setFetching] = useState(false);
  const [query, setQuery] = useState("");
  const [result, setResult] = useState();
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    if (query) {
      setFetching(true);
      setLoading(true);
      search(chainId as ChainId, query)
        .then((res) => {
          setResult(res);
          setFetching(false);
          setLoading(false);
        })
        .catch((searchError) => {
          logError({ searchError });
          setResult();
          setFetching(false);
          setLoading(false);
        });
    } else {
      setResult(undefined);
    }
  }, [chainId, query]);

  return (
    <SearchContext.Provider
      value={{ search: query, setSearch: setQuery, fetching, result, loading }}
    >
      {children}
    </SearchContext.Provider>
  );
};
