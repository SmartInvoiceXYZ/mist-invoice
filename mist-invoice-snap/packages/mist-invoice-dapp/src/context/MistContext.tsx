import React, { createContext, useContext, useEffect, useState } from 'react';

import { TransferType, UTXONote } from '../mist-sdk';
import {
  IncrementalMerkleTree,
  MerkleProof,
} from '@zk-kit/incremental-merkle-tree';
import { BytesLike, ethers, getBigInt, keccak256 } from 'ethers';
import { poseidon2 } from 'poseidon-lite';
import { randomBytes } from 'crypto';
import { useIndexedDB } from '../indexeddb';
import { logDebug, logError } from '../utils';
import { Web3Context } from './Web3Context';
import { connectSnap, getSnap, sendHello } from '@/snap/utils';
import {
  MetaMaskContext,
  MetamaskActions,
  MetamaskState,
} from './MetamaskContext';
import { DBConfig } from '@/dbconfig';

export type MistData = {
  merkleRoot?: BytesLike;
  clientRandom?: BytesLike;
  providerRandom?: BytesLike;
  clientKey?: BytesLike;
  providerKey?: BytesLike;
};

export type MistSecret = {
  merkleRoot?: bigint;
  providerHash: bigint;
  clientHash: bigint;
  encData: BytesLike[];
};

const utf8Encode = new TextEncoder();

export const SNARK_SCALAR_FIELD =
  21888242871839275222246405745257275088548364400416034343698204186575808495617n;
export const CLIENT_SIGNAL =
  getBigInt(keccak256(utf8Encode.encode('client'))) % SNARK_SCALAR_FIELD;
export const PROVIDER_SIGNAL =
  getBigInt(keccak256(utf8Encode.encode('provider'))) % SNARK_SCALAR_FIELD;
export const NULLIFYING_KEY =
  getBigInt(keccak256(utf8Encode.encode('nullifier'))) % SNARK_SCALAR_FIELD;

export type MistContextType = {
  data?: MistData;
  secret?: MistSecret;
  loading: boolean;
  amounts?: bigint[];
  receiver?: string;
  tokenAddress?: string;
  getClientProof: () => MerkleProof | undefined;
  handleSnapConnectClick: () => void;
  setAmounts: (amounts: bigint[]) => void;
  setReceiver: (receiver: string) => void;
  setTokenAddress: (tokenAddress: string) => void;
  state?: MetamaskState;
};

export const MistContext = createContext<MistContextType>({
  handleSnapConnectClick: () => undefined,
  getClientProof: () => undefined,
  loading: true,
  setAmounts: () => {},
  setReceiver: () => {},
  setTokenAddress: () => {},
});

export const MistContextProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const { getByID, add } = useIndexedDB(DBConfig.name, DBConfig);
  const { account } = useContext(Web3Context);
  const [data, setData] = useState<MistData>();
  const [secret, setSecret] = useState<MistSecret>();
  const [tree, setTree] = useState<IncrementalMerkleTree>();
  const [loading, setLoading] = useState(false);
  const [tokenAddress, setTokenAddress] = useState<string>('');
  const [receiver, setReceiver] = useState<string>('');
  const [amounts, setAmounts] = useState<bigint[]>([]);

  const providerAddress = process.env.MIST_PROVIDER_ADDRESS || '';

  // FROM MM SNAP TEMPLATE SITE
  const [state, dispatch] = useContext(MetaMaskContext);

  const handleSnapConnectClick = async () => {
    try {
      await connectSnap();
      const installedSnap = await getSnap();

      dispatch({
        type: MetamaskActions.SetInstalled,
        payload: installedSnap,
      });
    } catch (e) {
      console.error(e);
      dispatch({ type: MetamaskActions.SetError, payload: e });
    }
  };

  const handleSendHelloClick = async () => {
    try {
      await sendHello();
    } catch (e) {
      console.error(e);
      dispatch({ type: MetamaskActions.SetError, payload: e });
    }
  };
  //////////////////////////////////////////////////////////

  useEffect(() => {
    if (!account || !providerAddress || !data) return;
    const values = [
      [account, data.clientRandom],
      [providerAddress, data.providerRandom],
    ];

    const _tree = new IncrementalMerkleTree(poseidon2, 2, 0, 2, values);
    setTree(_tree);
    setData({
      ...data,
      merkleRoot: _tree.root,
    });
  }, [account, providerAddress, data]);

  useEffect(() => {
    const generate = async () => {
      const random = getBigInt(randomBytes(32).toString('hex'));
      // Contains bytes[] of encrypted note, clientKey and providerKey for each amount
      const encData = await Promise.all(
        amounts.map(async (amount, i) => {
          const note = new UTXONote({
            sender: account || '',
            receiver: receiver || '',
            amount,
            token: tokenAddress,
            identifier: BigInt(i),
            random,
            transferType: TransferType.Deposit,
            nullifyingKey: NULLIFYING_KEY,
          });
          const encryptedNote = await note.encrypt('goerli');
          return ethers.AbiCoder.defaultAbiCoder().encode(
            [
              'tuple(string encryptedData, string encryptedSenderKey, string encryptedReceiverKey)',
            ],
            [encryptedNote],
          );
        }),
      );
      const clientId = getBigInt(
        keccak256(
          ethers.AbiCoder.defaultAbiCoder().encode(['address'], [account]),
        ),
      );
      const providerId = getBigInt(
        keccak256(
          ethers.AbiCoder.defaultAbiCoder().encode(
            ['address'],
            [providerAddress],
          ),
        ),
      );
      const clientHash = poseidon2([
        poseidon2([clientId, NULLIFYING_KEY]),
        random,
      ]);
      const providerHash = poseidon2([
        poseidon2([providerId, NULLIFYING_KEY]),
        random,
      ]);

      const newSecret = {
        providerHash,
        clientHash,
        encData,
      };

      setSecret(newSecret);
      return secret;
    };

    async function genAddRes() {
      let res = await generate();
      add(res, account).then(() => {
        setSecret(res);
      });
    }

    if (account) {
      setLoading(true);
      try {
        getByID(account).then((res) => {
          if (res) {
            logDebug('Using existing data');
            setData(res);
          } else {
            logDebug('Generating new data');
            genAddRes();
          }
        });
      } catch (e) {
        logError(e);
      } finally {
        setLoading(false);
      }
    } else {
      setData(undefined);
    }
  }, [account, receiver, tokenAddress, amounts]);

  const getClientProof = () => {
    if (account && tree && data) {
      const proof = tree.createProof(
        tree.indexOf(poseidon2([getBigInt(account), CLIENT_SIGNAL])),
      );
      return proof;
    }
    return undefined;
  };

  return (
    <MistContext.Provider
      value={{
        data,
        secret,
        amounts,
        receiver,
        tokenAddress,
        loading,
        getClientProof,
        handleSnapConnectClick,
        setAmounts,
        setReceiver,
        setTokenAddress,
        state,
      }}
    >
      {children}
    </MistContext.Provider>
  );
};
