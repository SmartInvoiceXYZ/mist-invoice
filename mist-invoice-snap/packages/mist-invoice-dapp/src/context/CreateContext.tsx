import { AbiCoder, encodeBytes32String, TransactionResponse } from 'ethers';
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import {
  getResolvers,
  getWrappedNativeToken,
  // getInvoiceFactoryAddress,
  isValidLink,
  logError,
} from '../utils/helpers';
import { register } from '../utils/invoice';
// import { uploadMetadata } from "../utils";
import { Web3Context } from './Web3Context';

import { INSTANT_STEPS, ESCROW_STEPS } from '../utils/constants';

import { INVOICE_TYPES } from '../utils/constants';
import { useCreateInstant } from './create-hooks/useCreateInstant';
import { useCreateEscrow } from './create-hooks/useCreateEscrow';
import {
  MistInvoiceEscrowWrapper,
  MistInvoiceEscrowWrapper__factory,
} from '@/types';
import { uploadMetadata } from '@/utils';
import { MistContext } from './MistContext';

export type ProjectAgreement = {
  type: string;
  src: string;
  createdAt: string;
};

export type CreateContextType = {
  projectName?: string;
  projectDescription?: string;
  projectAgreement?: ProjectAgreement[];
  projectAgreementSource?: string;
  projectAgreementLinkType?: string;
  startDate?: number;
  endDate?: number;
  safetyValveDate?: number;
  clientAddress?: string;
  paymentAddress?: string;
  paymentDue?: bigint;
  paymentToken?: string;
  milestones?: string;
  termsAccepted?: boolean;
  arbitrationProvider?: string;
  payments?: bigint[];
  invoiceType?: string;
  deadline?: number;
  lateFee?: bigint;
  lateFeeInterval?: number;
  tx?: TransactionResponse;
  loading: boolean;
  currentStep: number;
  nextStepEnabled: boolean;
  goBackHandler: () => void;
  nextStepHandler: () => void;
  setProjectName: (projectName: string) => void;
  setProjectDescription: (projectDescription: string) => void;
  setProjectAgreement: (projectAgreement: ProjectAgreement[]) => void;
  setProjectAgreementSource: (projectAgreementSource: string) => void;
  setProjectAgreementLinkType: (projectAgreementLinkType: string) => void;
  setStartDate: (startDate: number) => void;
  setEndDate: (endDate: number) => void;
  setSafetyValveDate: (safetyValveDate: number) => void;
  setClientAddress: (clientAddress: string) => void;
  setPaymentAddress: (paymentAddress: string) => void;
  setPaymentDue: (paymentDue: bigint) => void;
  setPaymentToken: (paymentToken: string) => void;
  setMilestones: (milestones: string) => void;
  setTermsAccepted: (termsAccepted: boolean) => void;
  setArbitrationProvider: (arbitrationProvider: string) => void;
  setPayments: (payments: bigint[]) => void;
  setInvoiceType: (invoiceType: string) => void;
  setDeadline: (deadline: number) => void;
  setLateFee: (lateFee: bigint) => void;
  setLateFeeInterval: (lateFeeInterval: number) => void;
  createInvoice: () => Promise<void>;
};

export const CreateContext = createContext<CreateContextType>({
  loading: true,
  currentStep: -1,
  nextStepEnabled: false,
  goBackHandler: () => {},
  nextStepHandler: () => {},
  setProjectName: () => {},
  setProjectDescription: () => {},
  setProjectAgreement: () => {},
  setProjectAgreementSource: () => {},
  setProjectAgreementLinkType: () => {},
  setStartDate: () => {},
  setEndDate: () => {},
  setSafetyValveDate: () => {},
  setClientAddress: () => {},
  setPaymentAddress: () => {},
  setPaymentDue: () => {},
  setPaymentToken: () => {},
  setMilestones: () => {},
  setTermsAccepted: () => {},
  setArbitrationProvider: () => {},
  setPayments: () => {},
  setDeadline: () => {},
  setInvoiceType: () => {},
  setLateFee: () => {},
  setLateFeeInterval: () => {},
  createInvoice: async () => {},
});

export const CreateContextProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const { provider: rpcProvider, chainId } = useContext(Web3Context);
  const { data: mistData, secret: mistSecret } = useContext(MistContext);

  const [wrappedNativeToken, setWrappedNativeToken] = useState('');

  // project details
  const [invoiceType, setInvoiceType] = useState('');
  const [projectName, setProjectName] = useState('');
  const [projectDescription, setProjectDescription] = useState('');
  const [projectAgreementLinkType, setProjectAgreementLinkType] =
    useState('https');
  const [projectAgreementSource, setProjectAgreementSource] = useState('');
  const [projectAgreement, setProjectAgreement] = useState<ProjectAgreement[]>([
    {
      type: projectAgreementLinkType,
      src: projectAgreementSource,
      createdAt: Date.now().toString(),
    },
  ]);
  const [startDate, setStartDate] = useState<number>();
  const [endDate, setEndDate] = useState<number>();
  const [detailsHash, setDetailsHash] = useState(''); // ipfsHash for projectDetails

  // payment details
  const [clientAddress, setClientAddress] = useState('');
  const [paymentAddress, setPaymentAddress] = useState('');
  const [paymentDue, setPaymentDue] = useState(BigInt(0));
  const [paymentToken, setPaymentToken] = useState('');
  const [milestones, setMilestones] = useState('1');

  // escrow details
  const [safetyValveDate, setSafetyValveDate] = useState(0);
  const [termsAccepted, setTermsAccepted] = useState(false);
  const [arbitrationProvider, setArbitrationProvider] = useState('');
  const [requireVerification, setRequireVerification] = useState(true);

  // instant payment details
  const [deadline, setDeadline] = useState(0);
  const [lateFee, setLateFee] = useState(BigInt(0));
  const [lateFeeInterval, setLateFeeInterval] = useState(0);

  // payments chunks
  const [payments, setPayments] = useState([BigInt(0)]);
  const [tx, setTx] = useState<TransactionResponse>();
  const [loading, setLoading] = useState(false);

  // step handling
  const [currentStep, setStep] = useState(1);
  const [nextStepEnabled, setNextStepEnabled] = useState(false);
  const [allValid, setAllValid] = useState(false);

  useEffect(() => {
    if (!chainId) return;
    const resolvers = getResolvers(chainId);
    if (resolvers?.length > 0) setArbitrationProvider(resolvers[0]);
    const wrapNatToken = getWrappedNativeToken(chainId);
    setWrappedNativeToken(wrapNatToken);
    setPaymentToken(wrapNatToken);
  }, [chainId]);

  const { Escrow, Instant } = INVOICE_TYPES;

  // common for all invoice types
  const step1Valid = useMemo(() => {
    if (invoiceType === Escrow) {
      return (
        projectName !== undefined &&
        isValidLink(projectAgreementSource) &&
        safetyValveDate !== undefined &&
        safetyValveDate > new Date().getTime()
      );
    } else if (invoiceType === Instant) {
      return projectName !== undefined && isValidLink(projectAgreementSource);
    }
    return false;
  }, [
    projectName,
    projectAgreementSource,
    safetyValveDate,
    Escrow,
    Instant,
    invoiceType,
  ]);

  // handle invoice type
  const { escrowStep2Valid, escrowStep3Valid } = useCreateEscrow({
    step1Valid,
    allValid,
    clientAddress,
    paymentAddress,
    payments,
    paymentToken,
    paymentDue,
    milestones,
    termsAccepted,
    arbitrationProvider,
    setAllValid,
  });

  const { instantStep2Valid } = useCreateInstant({
    step1Valid,
    allValid,
    clientAddress,
    paymentAddress,
    paymentToken,
    paymentDue,
    milestones,
    setAllValid,
  });

  useEffect(() => {
    setProjectAgreement([
      {
        type: projectAgreementLinkType,
        src: projectAgreementSource,
        createdAt: Date.now().toString(),
      },
    ]);
  }, [projectAgreementSource, projectAgreementLinkType]);

  useEffect(() => {
    if (step1Valid && currentStep === 2) {
      uploadMetadata({
        projectName,
        projectDescription,
        projectAgreement,
        startDate: startDate ? Math.floor(startDate / 1000) : undefined,
        endDate: endDate ? Math.floor(endDate / 1000) : undefined,
      })
        .then(
          (hash) => setDetailsHash(hash),
          (ipfsError) => {
            logError({ ipfsError });
            setDetailsHash(
              '0xed70cef5a086e264386ab026edaaabb8f25c378f1e3a699a2c0a72a20d53c7fb',
            );
          },
        )
        .then(() => {
          console.log('uploadedMetadata', detailsHash);
        });
    }
  }, [
    step1Valid,
    currentStep,
    projectName,
    projectDescription,
    projectAgreement,
    projectAgreementLinkType,
    projectAgreementSource,
    startDate,
    endDate,
  ]);

  const encodeEscrowData = useCallback(
    (factoryAddress: string) => {
      const resolverType = 0; // 0 for individual, 1 for erc-792 arbitrator
      const type = encodeBytes32String(Escrow);

      const data = AbiCoder.defaultAbiCoder().encode(
        [
          'address',
          'uint8',
          'address',
          'address',
          'uint256',
          'bytes32',
          'address',
          'bool',
          'address',
        ],
        [
          clientAddress,
          resolverType,
          arbitrationProvider,
          paymentToken,
          Math.floor(safetyValveDate / 1000),
          detailsHash,
          wrappedNativeToken,
          requireVerification,
          factoryAddress,
        ],
      );

      return { type, data };
    },
    [
      clientAddress,
      arbitrationProvider,
      paymentToken,
      safetyValveDate,
      detailsHash,
      wrappedNativeToken,
      requireVerification,
      Escrow,
    ],
  );

  const encodeInstantData = useCallback(() => {
    const type = encodeBytes32String(Instant);
    const data = AbiCoder.prototype.encode(
      [
        'address',
        'address',
        'uint256',
        'bytes32',
        'address',
        'uint256',
        'uint256',
      ],
      [
        clientAddress,
        paymentToken,
        Math.floor(deadline / 1000),
        detailsHash,
        wrappedNativeToken,
        lateFee,
        lateFeeInterval,
      ],
    );

    return { type, data };
  }, [
    clientAddress,
    paymentToken,
    deadline,
    detailsHash,
    wrappedNativeToken,
    lateFee,
    lateFeeInterval,
    Instant,
  ]);

  const createInvoice = useCallback(async () => {
    let type;
    let data;

    console.log('createInvoice', chainId, allValid, detailsHash, mistSecret);

    if (
      chainId &&
      rpcProvider &&
      allValid &&
      detailsHash &&
      mistData?.merkleRoot &&
      mistData?.clientRandom &&
      mistData?.providerRandom &&
      mistData?.clientKey &&
      mistData?.providerKey &&
      mistSecret?.encData
    ) {
      setLoading(true);
      setTx(undefined);

      // const factoryAddress = getInvoiceFactoryAddress(chainId);
      const escrowWrapperAddress =
        process.env.NEXT_PUBLIC_MIST_INVOICE_ESCROW_WRAPPER;

      let paymentAmounts = [BigInt(0)];
      if (escrowWrapperAddress && invoiceType === Escrow) {
        const escrowInfo = encodeEscrowData(escrowWrapperAddress);
        type = escrowInfo.type;
        data = escrowInfo.data;
        paymentAmounts = payments;
        const wrapper =
          MistInvoiceEscrowWrapper__factory.connect(escrowWrapperAddress);
        const connected = wrapper.connect(await rpcProvider.getSigner());
        const txResult = await connected.createInvoice(
          {
            merkleRoot: mistData.merkleRoot,
            clientRandom: mistData.clientRandom,
            providerRandom: mistData.providerRandom,
            clientKey: mistData.clientKey,
            providerKey: mistData.providerKey,
          },
          paymentAmounts,
          data,
          type,
        );

        console.log('txResult', txResult);

        setTx(txResult);
      } else if (invoiceType === Instant) {
        // const instantInfo = encodeInstantData();
        // type = instantInfo.type;
        // data = instantInfo.data;
        // paymentAmounts = [paymentDue];
      }

      setLoading(false);
    }
  }, [
    chainId,
    rpcProvider,
    // clientAddress,
    paymentAddress,
    // arbitrationProvider,
    // paymentToken,
    payments,
    // safetyValveDate,
    // deadline,
    // lateFee,
    // lateFeeInterval,
    detailsHash,
    // requireVerification,
    // step1Valid,
    // escrowStep2Valid,
    // instantStep2Valid,
    // escrowStep3Valid,
    allValid,
    invoiceType,
    Escrow,
    Instant,
    encodeEscrowData,
    encodeInstantData,
    paymentDue,
  ]);

  useEffect(() => {
    if (invoiceType === Escrow) {
      if (currentStep === 1) {
        setNextStepEnabled(step1Valid);
      } else if (currentStep === 2) {
        setNextStepEnabled(escrowStep2Valid);
      } else if (currentStep === 3) {
        setNextStepEnabled(escrowStep3Valid);
      } else if (currentStep === 4) {
        setNextStepEnabled(true);
      } else {
        setNextStepEnabled(false);
      }
    }

    if (invoiceType === Instant) {
      if (currentStep === 1) {
        setNextStepEnabled(step1Valid);
      } else if (currentStep === 2) {
        setNextStepEnabled(instantStep2Valid);
      } else if (currentStep === 3) {
        setNextStepEnabled(true);
      }
    }
  }, [
    step1Valid,
    escrowStep2Valid,
    escrowStep3Valid,
    instantStep2Valid,
    currentStep,
    invoiceType,
    Escrow,
    Instant,
  ]);

  const goBackHandler = () => setStep((prevState: number) => prevState - 1);

  const nextStepHandler = useCallback(() => {
    let maxStep;
    switch (invoiceType) {
      case Escrow:
        maxStep = Object.keys(ESCROW_STEPS).length;
        break;
      case Instant:
        maxStep = Object.keys(INSTANT_STEPS).length;
        break;
      default:
        maxStep = 0;
    }
    if (nextStepEnabled) {
      if (currentStep === maxStep) return createInvoice();
      setStep((prevState: number) => prevState + 1);
    }
    return () => undefined;
  }, [
    nextStepEnabled,
    currentStep,
    invoiceType,
    createInvoice,
    Escrow,
    Instant,
  ]);

  return (
    <CreateContext.Provider
      value={{
        projectName,
        projectDescription,
        projectAgreement,
        projectAgreementSource,
        projectAgreementLinkType,
        startDate,
        endDate,
        safetyValveDate,
        clientAddress,
        paymentAddress,
        paymentDue,
        paymentToken,
        milestones,
        termsAccepted,
        arbitrationProvider,
        payments,
        tx,
        invoiceType,
        deadline,
        lateFee,
        lateFeeInterval,
        // setters
        setProjectName,
        setProjectDescription,
        setProjectAgreement,
        setProjectAgreementSource,
        setProjectAgreementLinkType,
        setStartDate,
        setEndDate,
        setSafetyValveDate,
        setClientAddress,
        setPaymentAddress,
        setPaymentDue,
        setPaymentToken,
        setMilestones,
        setTermsAccepted,
        setArbitrationProvider,
        setPayments,
        setInvoiceType,
        setDeadline,
        setLateFee,
        setLateFeeInterval,
        // creating invoice
        loading,
        createInvoice,
        // stepHandling
        currentStep,
        nextStepEnabled,
        goBackHandler,
        nextStepHandler,
      }}
    >
      {children}
    </CreateContext.Provider>
  );
};
