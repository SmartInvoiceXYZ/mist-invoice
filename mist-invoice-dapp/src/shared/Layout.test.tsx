import { render } from "@testing-library/react";
import React from "react";
import { Web3ContextProvider } from "../context";
import { Layout } from "./Layout";

jest.mock("@chakra-ui/react", () => ({
  ...jest.requireActual("@chakra-ui/react"),
  useBreakpointValue: () => "md",
}));
const mockPathname = jest.fn();
jest.mock("next/navigation", () => ({
  usePathname: mockPathname,
}));

describe("Layout", () => {
  it("should render", () => {
    // arrange
    mockPathname.mockReturnValue("/");

    // act
    const view = render(
      <Web3ContextProvider>
        <Layout />
      </Web3ContextProvider>
    );

    // assert
    expect(view.asFragment()).toMatchSnapshot();
  });
});
