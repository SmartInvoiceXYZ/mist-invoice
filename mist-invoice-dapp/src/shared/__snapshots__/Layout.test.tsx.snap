// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render 1`] = `
<DocumentFragment>
  <div
    class="css-149uwkc"
  >
    <div
      class="css-zzdl2l"
    >
      <div
        class="css-1hskriy"
      >
        <a
          href="/invoices"
        >
          <div
            class="css-11dqbsu"
          >
            <img
              alt="Smart Invoice"
              class="chakra-image css-12aolnz"
              src="normal.svg"
            />
          </div>
        </a>
      </div>
      <div
        class="css-1a9re40"
      >
        <a
          class="chakra-link css-0"
          href="/invoices"
        >
          Dashboard
        </a>
        <a
          class="chakra-link css-0"
          href="https://docs.smartinvoice.xyz"
          rel="noopener"
          target="_blank"
        >
          Documentation
        </a>
        <a
          class="chakra-link css-0"
          href="https://docs.smartinvoice.xyz/misc/get-support"
          rel="noopener"
          target="_blank"
        >
          Support
        </a>
      </div>
      <div
        class="css-uqd54k"
      />
      <div
        class="css-1mozfd5"
      >
        <button
          class="chakra-button css-czfkw6"
          type="button"
        >
          Dashboard
        </button>
        <a
          class="chakra-link css-3le2fk"
          href="https://docs.smartinvoice.xyz"
          rel="noopener"
          target="_blank"
        >
          <span
            class="chakra-button css-czfkw6"
          >
            Documentation
          </span>
        </a>
        <a
          class="chakra-link css-3le2fk"
          href="https://docs.smartinvoice.xyz/misc/get-support"
          rel="noopener"
          target="_blank"
        >
          <span
            class="chakra-button css-czfkw6"
          >
            Support
          </span>
        </a>
      </div>
    </div>
    <div
      class="css-90ijcl"
    />
    <div
      class="css-15cs0q7"
    >
      <div
        class="css-qymxha"
      >
        <a
          class="chakra-link css-0"
          href="/invoices"
        >
          <img
            alt="Smart Invoice"
            class="chakra-image css-lu3yos"
            src="white.svg"
          />
        </a>
        <div
          class="css-1a9re40"
        >
          <a
            class="chakra-link css-0"
            href="https://smartinvoice.xyz"
            rel="noopener"
            target="_blank"
          >
            Website
          </a>
          <a
            class="chakra-link css-0"
            href="https://docs.smartinvoice.xyz"
            rel="noopener"
            target="_blank"
          >
            Documentation
          </a>
          <a
            class="chakra-link css-0"
            href="https://docs.smartinvoice.xyz/misc/get-support"
            rel="noopener"
            target="_blank"
          >
            Support
          </a>
          <a
            class="chakra-link css-0"
            href="https://twitter.com/SmartInvoiceXYZ"
            rel="noopener"
            target="_blank"
          >
            Twitter
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
